File: 338CountingBits.cpp
/*
每次检查自己的上一个数的尾数为1的个数
dp[i] = dp[i-1] - 尾数1个数 + 1
*/

class Solution {
public:
    vector<int> countBits(int num) {
        vector<bitset<32>> v;
        v.reserve(num+1);
        for(int i = 0; i < num; i++) {
            bitset<32> bs32(i);
            v.push_back(bs32);
        }
        vector<int> dp(num+1, 0);
        
        for(int i = 1; i <= num; i++){
            if(v[i-1][0] == 0){
                dp[i] = dp[i-1] + 1;
            }
            int tail1 = 0;
            for(int j = 0; j < 32; j++){
                if(v[i-1][j] == 1){
                    tail1 += 1;
                }else{
                    break;
                }
            }
            dp[i] = dp[i-1] - tail1 + 1;
        }
        return dp;
    }
};

File: 338CountingBits.py
"""
		res[0] = 0
res[i] = 	res[i-1]+1    	# i为奇数
		res[i//2] 	# i为偶数
i为奇数时，i-1为偶数，其二进制组成最低位为0，所以i就是比i-1多了最低位这么一位1。而当i为偶数时，其二进制组成最低位为0，所以其1的位数就是除了最低位之外前面那一部分中1的位数，即是i/2中1的位数。		
"""
class Solution:
    def countBits(self, num):
        """
        :type num: int
        :rtype: List[int]
        """
        r = [0 for i in range(num+1)]
        for i in range(num+1):
            if i is 0:
                r[i] = 0
            else:
                # 偶数
                if i%2 is 0:
                    r[i] = r[i//2]
                # 奇数
                else:
                    r[i] = r[i-1] + 1  
        return r
